### DISCUSSION TITLE
Tournament Client Bracket UI Improvements

### DISCUSSION THREAD

**Discussion by Snowleopard-osu** [ppy/osu#13806] [Ideas]
The "Bracket Editor" section of the lazer tournament client can be a bit cumbersome to use, especially when making updates to an existing bracket.  The changes proposed below should help fix that and also result in `bracket.json` data being organized more sensibly.

#### Change how matches are assigned an ID behind the scenes.

Matches are currently given an ID number that correspond to their XY positioning, rather than creation time or chronological order. Here is an example of this with one of my tournaments. The ID number in `bracket.json` for each match is written in red.

![First Example](http://s.owo.lt/SmzXEA6b)

The current ID numbering does not match the order I created the matches, nor chronological succession, nor anything else useful to the user. Being able to have these correspond to creation order, or be user definable would make it a lot more useful. Compare the above numbering to this more traditional numbering system (Challonge):

![Challonge Example](http://s.owo.lt/mVCAO5lG)

It is also worth noting that despite lazer choosing to assign an ID based on XY positioning, the entries in `bracket.json` are still listed in the order they were created. The main reason why this change will be important is how it will tie into the next change proposals below.

#### Make the match ID visible to the user in the UI, and preferably user definable via the UI. 

The ID could be added to the current match details window as a textbox that a user could modify. Here is match 1 for reference:

![Lazer UI](http://s.owo.lt/t2gCAICz)

#### Expose the match 'progressions' in the UI and allow a user to modify them

Here is how this currently looks in `bracket.json`:

![progressions](http://s.owo.lt/52gQdbiJ) ![progressions loser match](http://s.owo.lt/i3hRHcXj)

This data is visible only via lines in the UI which overlap in `Bracket Editor` to the point you cannot tell what is going where, and when in the `Bracket` screen the lines for loser progressions aren't shown at all.

![Bracket Editor](http://s.owo.lt/eAlC2jJc) ![Bracket](http://s.owo.lt/szPYBBl2)

Compare this to how the data is displayed in Challonge:

![Challonge](http://s.owo.lt/jhObS61n)

By exposing the match ID to the user it would allow for setting up Lazer to match such a setup a lot more easily and also would provide a better viewing experience on stream when showing the `Bracket` screen.

#### Let the user indicate which slot a match progression corresponds to. Closely tied with the Third Change above

Using the above challonge example for match 24, we would want the loser of match 16 to fill into the top slot, and the loser of match 1 to fill into the bottom slot.

![Match 24](http://s.owo.lt/vGrzPJud)

There is currently no way to guarantee this, nor any way to know how it will work until you add scores to the conditional matches and hope for the best. From best I can tell, lazer just fills top to bottom based on whichever conditional match score is input first.  This is less than ideal and makes coordinating lobby title/what lobby slots players should be in/etc harder than it needs to be, since it is effected by conditional match scheduling.

### Closing Thoughts
While each of the 4 changes I listed would be nice on their own, combining all 4 together would result in a very noticeable improvement to how brackets in lazer are organized, modified, and displayed on stream. These changes would also improve the possibilities of automating creation of `bracket.json` via outside tools or data.  This is something I am very much interested in, but having these changes implemented first would make the job a lot easier.

**Comment by peppy** [comment]
You keep talking about IDs (ordering and display), but why do you need that in the first place? Is there something we can do to the editor to stop your need to edit the files manually, or is there something more to it? You're asking for changes here but I'm not seeing the rationale.

**Reply by Snowleopard-osu** [reply]
The rationale for having 'better' IDs is to facilitate the Third Change and the Fourth Change.  Using a match ID number is also a common and convenient way to reference a match. Compare "match # 9" to "the match between team X and team Y in the losers bracket of the quarter finals stage". 

Ideally there will never be a need to modify the files manually*, but exposing the match ID to the UI and allowing it to be used as described in the 3rd and 4th change will provide benefits in bracket management and display on stream.

\* Exception: using a outside tool to generate a bracket.json for a tournament defined elsewhere. A good example would be generating a bracket in Challonge and using a tool to generate a bracket.json via Challonge bracket data.

**Reply by peppy** [reply]
I see, so the intention is to also display the match ID/number on the streaming screen and use as a reference in commentary?

**Reply by Snowleopard-osu** [reply]
That would be ideal yes, the picture attached under Fourth Change should provide an example of how it can be useful in commentary and display on stream.

----

The Fourth Change in particular is also important for making lazer match a Challonge bracket and/or expected match setup in a referee client (such as Script Chan) where making the lobby title/player order match with Lazer is important.

__For example (3v3 match)__: If lazer thinks the match is Team A vs Team B then the lobby title should indicate that, additionally the players from Team A must be the red team and be in slots 1-3 to show up correctly on stream. Team B must be the blue team and be in slots 4-6.

In contrast, if lazer decides the match is actually Team B vs Team A you will need your lobby constructed to match this. ie Team B is the red team and is in slots 1-3, while Team A is the blue team and in slots 4-6.

When the match slot order is decided unpredictably by lazer it can be a real hassle to coordinate this or require the streamer(s) to frequently update the bracket to match some other authoritative source (such as a spreadsheet or Challonge)
